sklearn/metrics/pairwise.py

```diff
@@ -401,14 +407,26 @@ def _euclidean_distances(X, Y, X_norm_squared=None, Y_norm_squared=None, squared
         distances = -2 * safe_sparse_dot(X, Y.T, dense_output=True)
         distances += XX
         distances += YY
-    np.maximum(distances, 0, out=distances)
+
+    xp_zero = xp.asarray(0, device=device_, dtype=distances.dtype)
+    if _xp_method_has_out(xp_method=xp.maximum):
+        xp.maximum(distances, xp_zero, out=distances)
+    else:
+        distances = xp.maximum(distances, xp_zero)
 
     # Ensure that distances between vectors and themselves are set to 0.0.
     # This may not be the case due to floating point rounding errors.
     if X is Y:
-        np.fill_diagonal(distances, 0)
+        _add_to_diagonal(distances, 0, xp=xp, add_value=False)
 
-    return distances if squared else np.sqrt(distances, out=distances)
+    if squared:
+        return distances
+
+    if _xp_method_has_out(xp_method=xp.sqrt):
+        xp.sqrt(distances, out=distances)
+    else:
+        distances = xp.sqrt(distances)
```

Comment:
Maybe we could have a helper:

```python
def _use_out_if_possible(func, *args, out=None, **kwargs):
    try:
        func(*args, out=out, **kwargs)
    except TypeError:
        out = func(*args, **kwargs)
    return out
```

Note: we unfortunately have to resort to a blind try/except since we cannot rely on `"out" in inspect.signature(func).parameters` because numpy ufuncs only expose anonymous `*args` and `**kwargs` parameters.

Then use it as follows:

```python
distances = _use_out_if_possible(xp.sqrt, distances, out=distances)
```

Reply:
Alternatively only do it for numpy since the `out=` parameter will never be part of the array API spec so the semantics are not necessarily well defined.

```python
def _use_out_if_numpy(xp, func, *args, out=None, **kwargs):
    if _is_numpy_namespace(xp):
        func(*args, out=out, **kwargs)
    else:
        out = func(*args, **kwargs)
    return out
```

Then to use it:

```python
distances = _use_out_if_numpy(xp, xp.sqrt, distances, out=distances)
```

Reply:
I think I would vote for the latter, safer option.

Any opinions e.g. @betatim @lesteve @adrinjalali?

Reply:
I think we can start by adding array api support by following the standard only and discuss library memory optimizations later. For instance we could maintain an explicit allow-list of namespaces with `out=` kwarg support.

Reply:
Numpy is special because we don't want to introduce a memory efficiency regression for regular scikit-learn users.